crates/rust-analyzer/src/global_state.rs

```diff
@@ -87,7 +87,7 @@ pub(crate) struct GlobalState {
     pub(crate) flycheck_sender: Sender<flycheck::Message>,
     pub(crate) flycheck_receiver: Receiver<flycheck::Message>,
     pub(crate) last_flycheck_error: Option<String>,
-    pub(crate) diagnostics_received: bool,
+    pub(crate) diagnostics_received: FxHashMap<usize, bool>,
```

Comment:
I've added states based on the events in the main loop.

When I was changing it, I also think that the state should maybe track errors `last_flycheck_error` since currently, if there are multiple flychecks because of multiworkspace setup, the errors could be hidden because of the last check succeeding. But I didn't want to change further things since this is (hopefully) primarily about fixing a different bug, but I'll create a follow-up PR if there isn't an issue with that.